chore!: Upgrade wagtail to 5.x

asadali145 opened the pull request:
#### Pre-Flight checklist

- [x] Migrations
  - [x] Migration is backwards-compatible with current production code
- [x] Testing
  - [x] Code is tested
  - [x] Changes have been manually tested
- [x] Settings
  - [x] New settings are documented and present in `app.json`
  - [x] New settings have reasonable development defaults, if applicable

#### What are the relevant tickets?
Fixes https://github.com/mitodl/hq/issues/2865

#### What's this PR do?
Upgrades Wagtail version to 5.2.1(Latest)

- Disables Pylint as we are facing a string-based model references issue and I tried to resolve it but could not do that easily. We should do that in a separate issue.
- Updates Wagtail imports
- Updates wagtail-factories and wagtail-metadata to fix tests and bugs.
- Creates new migration required for the model changes after the Wagtail upgrade
- Moves all the data migrations to a single migration created at cms/0068.
- Fixes some migration dependencies to resolve migration issues caused by the renaming of PageRevision to Revision.
- Creates a migration at cms/0067 to populate the revision__content_type properly.
- Updates panels in cms/models.py as suggested by the Wagtail upgrade docs.
- Fixes broken tests.
- Removes model-admin for CourseTopic and creates snippet as suggested by Wagtail. Modeladmin is removed from wagtail but we can still use it in case we need it I think it is better to use snippets.
- Fixes courses app migrations to run before the PageRevision model rename migration.

#### How should this be manually tested?

- Checkout master branch
- Create some users, courses, programs, and certificates. Also, create sub-pages for the CoursePage, HomePage, and ProgramPage.
- Integrate xPro with edX (edX integration testing is optional. I have tested locally just to be sure)
- Create enrollments and coupons.
- Create Webinars
- Make sure that you have enough data to test all of the things.
- Now, run `docker-compose stop`
- Check out this branch
- Run `docker-compose build --no-cache`
- Run `docker-compose run --rm web ./manage.py makemigrations`
- Run `docker-compose run --rm web ./manage.py migrate`
- Run `docker-compose run --rm web ./manage.py configure_wagtail`
- Run `docker-compose run --rm web ./manage.py seed_data`
- Run `docker-compose up` and test all of the things that you did on master. Make sure that everything is working and there are no errors.
- Now run `docker-compose down` to remove the old db and run migrations and test all of the things again for a new instance.
- Make sure that the management commands are working fine (I have tested most of the related management commands and they look good to me)


#### Where should the reviewer start?
Review changes, set up your local xPro on the master branch, and then switch back to this branch.

#### Any background context you want to provide?
Pylint is disabled as it was having issues with the String-based model references. We will look into Pylint in a separate issue.

#### What GIF best describes this PR or how it makes you feel?
🥴

#### After Deployment
- [ ] Need to run `./manage.py rebuild_references_index` after the upgrade as suggested in the [docs](https://docs.wagtail.org/en/stable/releases/4.1.html#rebuild-references-index-management-command)

arslanashraf7 reviewed (changes requested):
I've left some comments and questions.

Although, In testing, things looked good on an existing instance. I'll test the fresh instance upgrade later.

asadali145 commented on cms/migrations/0060_webinarindexpage_webinarpage.py:
```diff
@@ -1,71 +1,12 @@
 
 import django.db.models.deletion
-import pytz
 import wagtailmetadata.models
 from django.db import migrations, models
-from wagtail.core.models import Page, PageRevision
 
 import cms.models
```
All of these imports are used. I haven't changed the import style to be consistent.

asadali145 commented on cms/migrations/0067_populate_revision_content_type.py:
```diff
@@ -0,0 +1,41 @@
+def populate_revision_content_type(apps, schema_editor):
+    ContentType = apps.get_model("contenttypes.ContentType")
+    Revision = apps.get_model("wagtailcore.Revision")
+    page_type = ContentType.objects.get(app_label="wagtailcore", model="page")
```
This is how Wagtail does in its migration and we are also doing it this way in our migrations.

asadali145 commented on cms/models.py:
```diff
@@ -1510,8 +1511,9 @@ class NewsAndEventsPage(Page):
         [("news_and_events", NewsAndEventsBlock())],
         blank=False,
         help_text="Add news and events updates to display in this section.",
+        use_json_field=True,
```
https://docs.wagtail.org/en/latest/releases/3.0.html#use-json-field-argument-added-to-streamfield

asadali145 commented on courses/wagtail_hooks.py:
```diff
@@ -0,0 +1,14 @@
+from wagtail.snippets.models import register_snippet
+from wagtail.snippets.views.snippets import DeleteView, SnippetViewSet
+
+from courses.models import CourseTopic
+
+
+class CourseTopicSnippet(SnippetViewSet):
+    """Snippet for CourseTopic"""
+
+    model = CourseTopic
+
+
+register_snippet(CourseTopicSnippet)
```
Yes, Moved from modeladmins to snippets. modeladmins are moved out of wagtail and wagtail recommends using snippets. I just wanted to avoid an extra dependency here.

arslanashraf7 replied:
I think the main reason initially was to keep the topics on the same levels as Snippets, But now the topics are moved under Snippets as a child.

See the previous look:
<img width="202" alt="image" src="https://github.com/mitodl/mitxpro/assets/34372316/2e9b1a44-3247-4806-b556-0ba77762a58c">

asadali145 replied:
Updated in https://github.com/mitodl/mitxpro/pull/2830/commits/6614b627ad2856d4a6f0511074b5865606e9a1bf

asadali145 commented on mitxpro/settings.py:
```diff
@@ -901,6 +900,9 @@
     description="The maximum number of cache entries for Wagtail images",
 )
+
+WAGTAILADMIN_BASE_URL = SITE_BASE_URL
```
It was added by Nathan and I just kept it. Wagtail has renamed its setting from `BASE_URL` to `WAGTAILADMIN_BASE_URL` and by reading the docs it looks like it should be pointing to `/cms` but I am not 100% clear on this one. We can remove this as well but its a good one to keep to avoid fallback options.

https://docs.wagtail.org/en/latest/reference/settings.html#wagtailadmin-base-url

Let me know what you think of this.

asadali145 commented on cms/migrations/0067_populate_revision_content_type.py:
```diff
@@ -0,0 +1,41 @@
+
+    run_before = [
+        ("wagtailcore", "0072_alter_revision_content_type_notnull"),
+    ]
```
I have removed the migration and things are looking good. I am not sure why we faced the issue maybe it got resolved after we reordered some of the migrations.

asadali145 commented on cms/models.py:
```diff
@@ -1856,9 +1863,10 @@ class ImageCarouselPage(CourseProgramChildPage):
     )
 
-    content_panels = Page.content_panels + [StreamFieldPanel("images")]
+    content_panels = [FieldPanel("title"), FieldPanel("images")]
```
Wagtail made some changes around the `title` field and added a new panel `TitleFieldPanel`. `TitleFieldPanel` has some issues when we disable the promote panels like in this case. It throws as an exception and that's the reason we needed to do this. Also, in this case, Page.content_panels only adds `title`. This slug issue might get resolved in newer versions as I saw it was logged somewhere.

<img width="1737" alt="Screen Shot 2023-12-05 at 5 43 51 PM" src="https://github.com/mitodl/mitxpro/assets/52656433/7132b122-fd3a-4772-9e00-c8f882bd5ff1">

asadali145 commented:
@arslanashraf7 I have rebased the PR and also removed the migration that we ported from Wagtail. Also, added some answers to your comments. Please let me know if you want to have a discussion.

arslanashraf7 reviewed (approved):
Looks good 👍 Thanks for making all the changes.
